### PR TITLE
Feat/notification settings

### PR DESCRIPTION
**Pull request by rugue**
# What does this PR do?
This PR implements endpoints for managing user notification settings:
* Added PATCH /api/v1/notification-settings Endpoint: Creates or updates user notification settings.
* Added GET /api/v1/settings/notification-settings/:user_id Endpoint: Retrieves user notification settings.
* Authentication: Ensures only authenticated users can access these endpoints.
* Validation: Implements input validation for notification settings data.
* Error Handling: Returns specific HTTP status codes and messages based on various scenarios.

# How should this be manually tested?
1. Login User:
   * Make a `POST` request to `/api/v1/auth/login` with a valid user's email and password.
   * Save the access token from the response.

2. Create/Update Notification Settings:
   * Make a `POST` request to `/api/v1/notification-settings` with the following:
     - Header: `Authorization: Bearer <access_token>`
     - Body:
       ```json
       {
         "mobile_push_notifications": "boolean",
         "email_notification_activity_in_workspace": "boolean",
         "email_notification_always_send_email_notifications": "boolean",
         "email_notification_email_digest": "boolean",
         "email_notification_announcement_and_update_emails": "boolean",
         "slack_notifications_activity_on_your_workspace": "boolean",
         "slack_notifications_always_send_email_notifications": "boolean",
         "slack_notifications_announcement_and_update_emails": "boolean"
       }
       ```
   * Verify the response contains the created/updated notification settings.

3. **Get Notification Settings:**
   * Make a `GET` request to `/api/v1/settings/notification-settings/user@example.com` with the access token.
   * Verify the response contains the user's notification settings.

4. Edge Cases:
   * Unauthenticated: Test without a token to ensure the endpoint returns a `401 Unauthorized` status.
   * Invalid User: Test with a non-existent user_id to ensure the endpoint returns a `404 Not Found` status.
   * Invalid Input: Test with invalid data to ensure proper validation errors are returned.

# Checklist:
- [x] Implemented `PATCH /api/v1/settings/notification-settings` endpoint in `notification-settings.controller.ts`.
- [x] Implemented `GET /api/v1/settings/notification-settings/:user_id` endpoint in `notification-settings.controller.ts`.
- [x] Added `createOrUpdate` and `findByUserId` methods in the `NotificationSettingsService`.
- [x] Implemented input validation using DTO.
- [x] Tested endpoint functionality to ensure it correctly handles various scenarios.
- [x] Added appropriate HTTP status codes and response messages for edge cases.
- [x] Added unit tests for the `NotificationSettingsService`.
- [x] Updated API documentation to reflect the new endpoints.

# Response Examples
**Success (PATCH/GET):**

```json
{
  "status": "success",
  "message": "Notification preferences updated successfully",
  "status_code": 200,
  "data": {
    "email_notification_activity_in_workspace": true,
    "email_notification_always_send_email_notifications": false,
    "email_notification_email_digest": true,
    "email_notification_announcement_and_update_emails": true,
    "slack_notifications_activity_on_your_workspace": false,
    "slack_notifications_always_send_email_notifications": true,
    "slack_notifications_announcement_and_update_emails": false
  }
}
```

**Error (Unauthorized):**

```json
{
  "message": "Unauthorized",
  "status_code": 401
}
```

**Error (Not Found):**

```json
{
  "status": "error",
  "message": "Notification settings not found",
  "status_code": 404
}
```

**Error (Validation):**

```json
{
  "status": "error",
  "message": [
    "user_id must be a valid email",
    "email_notifications must be a boolean",
    "push_notifications must be a boolean",
    "sms_notifications must be a boolean"
  ],
  "status_code": 400
}
```

# Link to Issue
Link to this issue: [#[FEAT] Implement Backend for Notification Settings #20](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/20)

### CODE REVIEW

File: src/modules/notification-settings/notification-settings.service.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      }
+      Object.assign(settings, notificationSettingsDto);
+      await this.notificationSettingsRepository.save(settings);
+      return { ...notificationSettingsDto };
+    } catch (error) {
+      this.logger.error('Error updating notification settings:', error);
+      if (error instanceof NotFoundException) {
+        throw error;
+      }
+      throw new BadRequestException('Invalid data in the request body');
```
Comment (Cobidex) [inline src/modules/notification-settings/notification-settings.service.ts:75]: invalid data in the request body should. It get to your service layer. Validation should address that.

File: src/modules/notification-settings/dto/notification-settings-response.dto.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsBoolean, IsEmail, IsNotEmpty } from 'class-validator';
```
Comment (CFCIfe) [inline src/modules/notification-settings/dto/notification-settings-response.dto.ts:2]: Remove this line. Its an unused import

File: src/modules/notification-settings/notification-settings.service.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        email_notification_email_digest: true,
+        email_notification_announcement_and_update_emails: true,
+        slack_notifications_activity_on_your_workspace: false,
+        slack_notifications_always_send_email_notifications: true,
+        slack_notifications_announcement_and_update_emails: false,
+      };
+
+      let settings = await this.notificationSettingsRepository.findOne({ where: { user_id } });
+      if (!settings) {
+        settings = await this.create(defaultNotificationSettings, user_id);
```
Comment (CFCIfe) [inline src/modules/notification-settings/notification-settings.service.ts:49]:
```suggestion
settings = await this.createNotificationSettings(defaultNotificationSettings, user_id);
```

File: src/modules/notification-settings/notification-settings.service.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      }
+
+      return settings;
+    } catch (error) {
+      this.logger.error('Error finding notification settings by user ID:', error);
+      throw new BadRequestException(error?.message || 'Failed to fetch notification settings');
+    }
```
Comment (CFCIfe) [inline src/modules/notification-settings/notification-settings.service.ts:56]: Your catch block can also catch dB errors. So this would send a false response. Reevaluate the badRequest and InternalServerExceptions

File: src/modules/notification-settings/dto/notification-settings-response.dto.ts (hunk 2 above)
Comment (rugue) [reply src/modules/notification-settings/dto/notification-settings-response.dto.ts:2]: done

File: src/modules/notification-settings/notification-settings.service.ts (hunk 1 above)
Comment (CFCIfe) [inline src/modules/notification-settings/notification-settings.service.ts:75]: Your catch block can also catch dB errors. So this would send a false response. Reevaluate the badRequest and InternalServerExceptions

**Review by CFCIfe — COMMENTED** [review]
Resolve the conversation, and push new changes

**Review by Heba-WebDev — APPROVED** [review]

**Review by davidolanrewaju — DISMISSED** [review]

**Review by davidolanrewaju — APPROVED** [review]

**Review by Heba-WebDev — APPROVED** [review]

**Review by TheCodeGhinux — APPROVED** [review]
Goood 🚀
